toolchain/check/name_component.cpp

```diff
@@ -43,11 +43,24 @@ auto PopNameComponent(Context& context, SemIR::InstId return_slot_pattern_id)
     implicit_param_patterns_id = SemIR::InstBlockId::Invalid;
   }
 
-  auto call_params_id =
-      CalleePatternMatch(context, *implicit_param_patterns_id,
-                         *param_patterns_id, return_slot_pattern_id);
+  Parse::NodeId name_loc_id = Parse::NodeId::Invalid;
+  auto name_id = SemIR::NameId::Invalid;
+  auto call_params_id = SemIR::InstBlockId::Invalid;
+  auto pattern_block_id = SemIR::InstBlockId::Invalid;
+  if (param_patterns_id->is_valid() || implicit_param_patterns_id->is_valid()) {
+    std::tie(name_loc_id, name_id) =
+        context.node_stack()
+            .PopWithNodeId<Parse::NodeKind::IdentifierNameBeforeParams>();
+    call_params_id =
+        CalleePatternMatch(context, *implicit_param_patterns_id,
+                           *param_patterns_id, return_slot_pattern_id);
+    pattern_block_id = context.pattern_block_stack().Pop();
+  } else {
+    std::tie(name_loc_id, name_id) =
+        context.node_stack()
+            .PopWithNodeId<Parse::NodeKind::IdentifierNameNotBeforeParams>();
+  }
```

Comment:
```suggestion
  auto call_params_id = SemIR::InstBlockId::Invalid;
  auto pattern_block_id = SemIR::InstBlockId::Invalid;
  if (param_patterns_id->is_valid() || implicit_param_patterns_id->is_valid()) {
    call_params_id =
        CalleePatternMatch(context, *implicit_param_patterns_id,
                           *param_patterns_id, return_slot_pattern_id);
    pattern_block_id = context.pattern_block_stack().Pop();
  }

  auto [name_loc_id, name_id] = context.node_stack().PopWithNodeId<Parse::NodeCategory::AnyNonExprIdentifierName>();
```

Note this doesn't validate the node kind match, but thought I'd suggest it since it'd get you closer to the prior flow if you want. This would also be a `NodeCategory` use per my other comment.

Reply:
Oh, thanks, that's much better!